feat: add typings for hooks

bradenmacdonald opened the pull request:
## Description

This PR adds typescript types for the various hooks that Paragon provides: `useWindowSize`, `useToggle`, `useArrowKeyNavigation`, `useIndexOfLastVisibleChild`, and `useIsVisible`. There should be no changes other than to the typings.

### Deploy Preview

https://deploy-preview-3125--paragon-openedx-v22.netlify.app/

## Merge Checklist

n/a - doesn't apply to this type of PR.

## Post-merge Checklist

* [ ] Verify your changes were released to [NPM](https://www.npmjs.com/package/@openedx/paragon) at the expected version.
* [ ] If you'd like, [share](https://github.com/openedx/paragon/discussions/new?category=show-and-tell) your contribution in [#show-and-tell](https://github.com/openedx/paragon/discussions/categories/show-and-tell).
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

bradenmacdonald committed:
feat: add typings for hooks

openedx-webhooks commented:
Thanks for the pull request, @bradenmacdonald! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

bradenmacdonald committed:
fix: wrong return type specified in www Navbar

bradenmacdonald committed:
fix: useIsVisible() can take any element, not just <div>

bradenmacdonald commented:
OK, this is very weird. Without any the other changes, merely renaming `useIsVisible.jsx` to `useIsVisible.tsx` totally breaks the build of the Paragon docs site (in dev mode, if I restart it, I cannot access the site). But the error is very obscure.

adamstankiewicz commented on src/hooks/tests/useIndexOfLastVisibleChild.test.tsx:
```diff
@@ -12,7 +12,7 @@ window.ResizeObserver = window.ResizeObserver
 function TestComponent() {
-  const [containerElementRef, setContainerElementRef] = React.useState(null);
+  const [containerElementRef, setContainerElementRef] = React.useState<HTMLDivElement | null>(null);
   const overflowElementRef = React.useRef(null);
```
nit/question: should both refs get the same type with `HTMLDivElement`, given both refs are applied to `divs` below?

bradenmacdonald replied:
Done: a3fea9e291c5b3fc28ccd4be154d5d5695cd8009

Note that changing it to `useRef` instead of `useState` for consistency seems to make one of the tests fail, so I've just left one as `useState` and one as `useRef` for now, but at least they're both `HTMLDivElement` refs.

adamstankiewicz commented on src/hooks/tests/useToggle.test.tsx:
```diff
@@ -19,7 +20,7 @@ const resetHandlerMocks = () => {
 };
 
 // eslint-disable-next-line react/prop-types
```
nit/unrelated suggestion: could opt to disable prop-types rule for `*.test.*` files so we don't need to have these eslint disable comments throughout test files.

bradenmacdonald replied:
Sure, but I'm gonna leave that out of this PR for now.

adamstankiewicz commented on src/hooks/useArrowKeyNavigation.tsx:
```diff
@@ -4,13 +4,21 @@ import { useRef, useEffect } from 'react';
  */
 
-function handleEnter({ event, currentIndex, activeElement }) {
+function handleEnter(
+  { event, currentIndex, activeElement }: { event: KeyboardEvent, currentIndex: number, activeElement: HTMLElement },
+) {
```
curious/suggestion: would it improve readability/understandability of these types if these function args and the return data structure were abstracted into an `interface`? E.g., something like the following:

```tsx
interface HandleEnterOptions {
  event: KeyboardEvent;
  currentIndex: number;
  activeElement: HTMLElement;
}

function handleEnter({ event, currentIndex, activeElement }: HandleEnterOptions) { ... }
```

Similar question for `handleArrowKey` below

bradenmacdonald replied:
Sure :) a3fea9e291c5b3fc28ccd4be154d5d5695cd8009

adamstankiewicz commented on src/hooks/useIndexOfLastVisibleChild.tsx:
```diff
@@ -5,25 +5,32 @@ import { useLayoutEffect, useState } from 'react';
  */
-const useIndexOfLastVisibleChild = (containerElementRef, overflowElementRef) => {
+const useIndexOfLastVisibleChild = (
+  containerElementRef: Element | null,
+  overflowElementRef: Element | null,
```
nit/clarification: should these `Element` be `HTMLElement`?

bradenmacdonald replied:
Yeah - done in a3fea9e291c5b3fc28ccd4be154d5d5695cd8009

adamstankiewicz commented on src/hooks/useIsVisible.tsx:
```diff
@@ -1,7 +1,10 @@
+  isVisible: boolean,
+  sentinelRef: React.MutableRefObject<HTMLDivElement | null>,
+] => {
+  const sentinelRef = useRef<HTMLDivElement | null>(null);
```
[clarification] Will this `sentinelRef` _always_ be applied to a `Div` element? Might it be used with non-`div` elements, too?

bradenmacdonald replied:
Good point, changed to `HTMLElement` a3fea9e291c5b3fc28ccd4be154d5d5695cd8009

mphilbrick211 commented:
@bradenmacdonald just checking in to see if this is still in progress!

bradenmacdonald commented:
@mphilbrick211 Thanks for the reminder. I lost momentum on this but now that I have @adamstankiewicz's feedback I would like to pick it up again soon!

bradenmacdonald commented:
Update: I tried to move this forward, but I'm blocked on some strange webpack bug where it is confusing `useToggle.mdx` and `useToggle.tsx`, returning the MDX content when I import the TypeScript file. I believe the only way to fix it is likely upgrading gatsby.

brian-smith-tcril commented:
> Update: I tried to move this forward, but I'm blocked on some strange webpack bug where it is confusing `useToggle.mdx` and `useToggle.tsx`, returning the MDX content when I import the TypeScript file. I believe the only way to fix it is likely upgrading gatsby.

I pulled the branch and got the docs site to build/run locally with a less than ideal workaround. I added `Doc` to the filenames of all the `.mdx` files in the `hooks` directory. Unfortunately that led to the url for those pages also having `doc` in them (example: `http://localhost:8000/components/hooks/usetoggledoc/`)

I'll do a little more digging to see if I can come up with a better option.

brian-smith-tcril commented:
I tried another thing which seems to work (I haven't clicked around all the docs site pages yet)

```diff
diff --git a/www/gatsby-config.js b/www/gatsby-config.js
index e3976a0b2c..c0ae5209a2 100644
--- a/www/gatsby-config.js
+++ b/www/gatsby-config.js
@@ -34,6 +34,7 @@ const plugins = [
     },
   },
   'gatsby-plugin-react-helmet',
+  'gatsby-plugin-typescript',
   {
     resolve: 'gatsby-plugin-manifest',
     options: {

```

bradenmacdonald committed:
fix: gatsby site wasn't building correctly after some .jsx -> .tsx changes

bradenmacdonald committed:
refactor: cleanups suggested by Adam

bradenmacdonald commented:
@adamstankiewicz Sorry for the long delay in addressing your review, and thanks SO much to @brian-smith-tcril for helping me figure out the very obscure bug that had me stumped.

This is ready for another review !

brian-smith-tcril reviewed (approved):
LGTM!

brian-smith-tcril commented on www/gatsby-config.js:
```diff
@@ -34,6 +34,7 @@ const plugins = [
     },
   },
   'gatsby-plugin-react-helmet',
+  'gatsby-plugin-typescript',
```
Note for reviewers: you may have noticed there are no `package.json` changes adding this, that is because `gatsby-plugin-typescript` is already a dependency of `gatsby` (v4 at least - I assume later versions as well) https://www.npmjs.com/package/gatsby/v/4.14.1?activeTab=dependencies
